### PR TITLE
[#135666429] Batch Process - Bulk Upsert

### PR DESCRIPTION
**Pull request by mattsmee**
https://www.pivotaltracker.com/story/show/135666429
- added `bulk_upsert` to couch
- added `bulk_upsert` to mysql and memory store
- fixed a bug in existing couch `upsert` function

### CODE REVIEW

File: lib/couchdb/index.js [hunk 1]
```diff
@@ -83,6 +83,29 @@ const upsert = R.curry((couch, bucket, keys, doc) =>
 )(doc)))
 
 
+const bulk_upsert = R.curry((couch, bucket, keys, docs) =>
+  Bluebird.map(docs, (doc) =>
+    Bluebird.resolve(keys)
+    .then(R.flip(R.pick)(doc))
+    .then(R.ifElse(
```
Comment (jonajonlee) [inline lib/couchdb/index.js:90]: looks like the H&M for this anonymous function will be 

`anon :: Predicates -> doc | [ doc ]`

Then the next anon function on line 99 will take the head of the passed value and couchify.

I think R.head might run on a single document that's returned in the above's `R.always(doc)` which would be bad right?

File: lib/couchdb/index.js [hunk 2]
```diff
@@ -83,6 +83,29 @@ const upsert = R.curry((couch, bucket, keys, doc) =>
...
+      R.isEmpty
+    , R.always(doc)
```
Comment (jonajonlee) [inline lib/couchdb/index.js:92]: So I think, maybe this should be `R.always({})` ?

File: lib/couchdb/index.js (hunk 1 above)
Comment (jonajonlee) [reply lib/couchdb/index.js:90]: Nvm, it seems to work because `R.merge(undefined, {a:1})` will return `{a:1}`

i.e. https://goo.gl/bu7NEJ

**Review by jonajonlee — APPROVED** [review]
:shipit: 


Ignore my comments, I've verified that those work.

**Review by scull7 — APPROVED** [review]

**Pull request merged**